### PR TITLE
Cluster provisioning - app sre bot automation

### PR DESCRIPTION
**Pull request by patjlm**
[APPSRE-8784](https://issues.redhat.com/browse/APPSRE-8784)

automatically create and register app-sre serviceaccounts on reachable clusters

### CODE REVIEW

**Review by geoberle — COMMENTED** [review]
looks good. i will have another look once it is out of draft

File: reconcile/openshift_clusterrolebindings.py [hunk 1]
```diff
@@ -159,6 +159,7 @@ def run(dry_run, thread_pool_size=10, internal=None, use_jump_host=True, defer=N
         cluster_info
         for cluster_info in queries.get_clusters()
         if cluster_info.get("managedClusterRoles")
+        and cluster_info.get("automationToken")
```
[the remark this answers is not in the document]
Comment (patjlm) [reply reconcile/openshift_clusterrolebindings.py:162]: yes it is, i'll do a separate ticket/MR for this.

File: reconcile/openshift_cluster_bots.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    try:
+        res = urllib.request.urlopen(url, timeout=10)
+        return res is not None and res.getcode() == 200
+    except URLError as e:
+        logging.debug(f"[{cluster.name}] API URL unreachable: {e.reason}")
+        return False
+
+
+def vault_secret(cluster: ClusterV1, cluster_admin: bool = False) -> dict[str, str]:
+    secret_key = f"app-sre/creds/kube-configs/{cluster.name}"
```
[the remark this answers is not in the document]
Comment (patjlm) [reply reconcile/openshift_cluster_bots.py]: extracted that into integration parameters

File: reconcile/openshift_clusterrolebindings.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (patjlm) [reply reconcile/openshift_clusterrolebindings.py:162]: https://github.com/app-sre/qontract-reconcile/pull/4019

File: reconcile/openshift_cluster_bots.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    submit_mr(clusters, gitlab_project_id, sqs_or_gitlab)
+
+
+def filter_clusters(clusters: list[ClusterV1]) -> list[ClusterV1]:
+    return [
+        cluster
+        for cluster in clusters
+        if integration_is_enabled(QONTRACT_INTEGRATION, cluster)
+        and cluster.ocm is not None
+        and cluster_misses_bot_tokens(cluster)
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/openshift_cluster_bots.py:270]: i misplaced my question, sry :)  . i meant the line below about `cluster_is_reachable`

File: reconcile/openshift_clusterrolebindings.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (patjlm) [reply reconcile/openshift_clusterrolebindings.py]: removed the corresponding commit here

File: reconcile/openshift_cluster_bots.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (patjlm) [reply reconcile/openshift_cluster_bots.py:270]: yes this checks if the API is reachable from the integration

**Review by geoberle — APPROVED** [review]

**Pull request merged**